src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AdminControllerTests.cs

```diff
@@ -0,0 +1,218 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using DevAdventCalendarCompetition.Controllers;
+using DevAdventCalendarCompetition.Models.Test;
+using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Resources;
+using DevAdventCalendarCompetition.Services.Interfaces;
+using DevAdventCalendarCompetition.Services.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Moq;
+using Xunit;
+using static DevAdventCalendarCompetition.Tests.TestHelper;
+
+namespace DevAdventCalendarCompetition.Tests.UnitTests.Controllers
+{
+    public class AdminControllerTests
+    {
+        private readonly Mock<ITestService> _testServiceMock;
+        private readonly Mock<IAdminService> _adminServiceMock;
+
+        public AdminControllerTests()
+        {
+           this._testServiceMock = new Mock<ITestService>();
+           this._adminServiceMock = new Mock<IAdminService>();
+        }
+
+        [Fact]
+        public void Index_ReturnsAViewResultWithAListOfTestDto()
+        {
+            // Arrange
+            var testList = new List<TestDto>()
+            {
+                GetTestDto()
+            };
+
+            this._adminServiceMock.Setup(x => x.GetAllTests()).Returns(testList);
+
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.Index();
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            var model = Assert.IsAssignableFrom<List<TestDto>>(viewResult.ViewData.Model);
+            Assert.Single(model);
+        }
+
+        [Fact]
+        public void AddTest_ModelStateIsInvalid_ReturnsViewWithInvalidModel()
+        {
+            // Arrange
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+            controller.ModelState.AddModelError("Number", "Required");
```

Comment:
why do we need to do that? it is not checking the logic from controller

we should run _controller.AddTest_ with the invalid object and expect that _controller.ModelState_ is invalid and contains specific error (from resources probably)

Reply:
Validation occurs outside the controller, as part of the MVC framework, so the scenario that you described won't work while unit testing.